pkg/auth/oauth/registry/timeoutvalidator.go

```diff
@@ -0,0 +1,278 @@
+		if client.AccessTokenInactivityTimeoutSeconds == nil {
+			continue
+		}
+		if *client.AccessTokenInactivityTimeoutSeconds == 0 {
+			continue
+		}
+		if *client.AccessTokenInactivityTimeoutSeconds < timeout {
+			timeout = *client.AccessTokenInactivityTimeoutSeconds
+			clientType = "OAuthClient"
+			clientName = client.Name
+		}
+	}
+	previousFlushTimeout := a.flushTimeout
+	_, a.flushTimeout, a.safetyMargin = a.computeTimeouts(timeout)
+	if a.flushTimeout != previousFlushTimeout {
+		glog.V(5).Infof("%s %s updates timeouts to flushTimeout=%d safetyMargin=%d",
+			clientType, clientName, int(a.flushTimeout.Seconds()), int(a.safetyMargin.Seconds()))
+	}
+}
+
+func (a *TimeoutValidator) update(td tokenData) error {
+	// Obtain the timeout interval for this client
+	delta := a.clientTimeout(td.token.ClientName)
+	// if delta is 0 it means the OAuthClient has been changed to the
+	// no-timeout value. In this case we set newTimeout also to 0 so
+	// that the token will no longer timeout once updated.
+	newTimeout := int32(0)
+	if delta > 0 {
+		// TimeoutsIn is the number of seconds since creation:
+		// TimeoutsIn = Seen(Time) - CreationTimestamp(Time) + delta(Duration)
+		newTimeout = int32((td.seen.Sub(td.token.CreationTimestamp.Time) + delta) / time.Second)
+	}
+	// We need to get the token again here because it may have changed in the
+	// DB and we need to verify it is still worth updating
+	token, err := a.tokens.Get(td.token.Name, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if newTimeout != 0 && token.TimeoutsIn >= newTimeout {
+		// if the token was already updated with a higher or equal timeout we
+		// do not have anything to do
+		return nil
+	}
+	token.TimeoutsIn = newTimeout
+	_, err = a.tokens.Update(token)
+	return err
+}
+
+func (a *TimeoutValidator) flush(flushHorizon time.Time) {
+	glog.V(5).Infof("Flushing tokens timing out before %s", flushHorizon)
+
+	// grab all tokens that need to be update in this flush interval
+	// and remove them from the stored data, they either flush now or never
+	tokenList := a.data.LessThan(flushHorizon.Unix(), true)
+
+	var retryList []tokenData
+	flushedTokens := 0
+
+	for _, item := range tokenList {
+		td := item.(*tokenData)
+		err := a.update(*td)
+		switch {
+		case err == nil:
+			flushedTokens++
+		case apierrors.IsConflict(err) || apierrors.IsServerTimeout(err):
+			glog.V(5).Infof("Token update deferred for token=%q, retriable error: %v",
+				td.token.Name, err)
+			retryList = append(retryList, *td)
+		default:
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
+				td.token.UserName, td.token.ClientName, td.token.Scopes, err)
+		}
+	}
+
+	// we try once more and if it still fails we stop trying here and defer
+	// to a future regular update if the token is used again
+	for _, td := range retryList {
+		err := a.update(td)
+		if err != nil {
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
```

Comment:
here

Reply:
neither here, is the err what you want me drop ? (doing that anyway)